packages/vite/src/node/plugins/assetImportMetaUrl.ts

```diff
@@ -62,16 +65,31 @@ export function assetImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
           }
 
           const url = rawUrl.slice(1, -1)
-          const file = path.resolve(path.dirname(id), url)
-          // Get final asset URL. Catch error if the file does not exist,
-          // in which we can resort to the initial URL and let it resolve in runtime
-          const builtUrl = await fileToUrl(file, config, this).catch(() => {
+          assetResolver ??= config.createResolver({
+            root: config.publicDir,
+            extensions: [],
+            mainFields: [],
+            tryIndex: false,
+            preferRelative: true
+          })
+          const resolved = await assetResolver(url, id)
+          if (!resolved) {
             const rawExp = code.slice(index, index + exp.length)
             config.logger.warnOnce(
               `\n${rawExp} doesn't exist at build time, it will remain unchanged to be resolved at runtime`
             )
-            return url
-          })
+          }
+          // Get final asset URL. If the file does not exist,
+          // we fall back to the initial URL and let it resolve in runtime
+          const builtUrl = resolved
+            ? await fileToUrl(
+                isParentDirectory(normalizePath(config.publicDir), resolved)
+                  ? url
+                  : resolved,
+                config,
+                this
+              )
+            : url
```

Comment:
I think using an `if` block would make this easier to read. Also have a question with the `config.publicDir` mentioned above.